### ISSUE TITLE
Definition of push_front() doesn't sound right.

### ISSUE THREAD

**Issue by vgurevich** [p4lang/p4-spec#225]
From the spec:

> ```hs.push_front(int count)```: shift "right" by ```count```. The
>   first ```count``` elements become invalid. The last ```count```
>   elements in the stack are discarded. The ```hs.nextIndex``` counter
>   is incremented by ```count```.

I think that the first statement (about "shift right") is correct, but I have trouble visualizing the rest.

**Comment by jafingerhut** [comment]
The description makes sense to me, but perhaps it would help if there were pseudocode to describe these operations as well?

```
~ Begin P4Pseudo
void hs.push_front(int count) {
    for (int i=hs.size-1; i >= 0; i--) {
        if (i >= count) {
            hs[i] = hs[i-count];
        } else {
            hs[i].setInvalid();
        }
    }
    hs.nextIndex = hs.nextIndex + count;
    // TBD: Should nextIndex be adjusted to hs.size if it goes past there?  Seems like a good idea.
    if (hs.nextIndex > hs.size) hs.nextIndex = hs.size;
    // Note: hs.last, hs.next, and hs.lastIndex adjust with hs.nextIndex
}

void hs.pop_front(int count) {
    for (int i=0; i < hs.size; i++) {
        if (i < hs.size - count) {
            hs[i] = hs[i+count];
        } else {
            hs[i].setInvalid();
        }
    }
    hs.nextIndex = hs.nextIndex - count;
    // TBD: Should nextIndex be adjusted to 0 if it goes past there?  Seems like a good idea.
    if (hs.nextIndex < 0) hs.nextIndex = 0;
    // Note: hs.last, hs.next, and hs.lastIndex adjust with hs.nextIndex
}
~ End P4Pseudo
```

**Comment by jnfoster** [comment]
It makes sense to me. 

And I think adding the pseudocode. To answer the TBDs: your semantics for lastIndex looks consistent with what we do elsewhere.

**Comment by jafingerhut** [comment]
Should I create PR for adding the pseudocode?  I will take out the TBD comments if so.

**Comment by mbudiu-vmw** [comment]
I think that a drawing would be much more useful. But pseudo-code is necessary, especially for the corner cases.

**Comment by jafingerhut** [comment]
@vgurevich Does the pseudocode clarify anything for you?  A slightly modified version of what appears above is now committed to the spec.  You can see the committed diff here: https://github.com/p4lang/p4-spec/pull/229/files

**Comment by vgurevich** [comment]
@jafingerhut Looks right (with nextindex adjustment). Thanks a lot.

**Comment by jnfoster** [comment]
Closed by #229
